src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -316,9 +336,92 @@ private string FormatEnumValueToString(string value, JavaScriptEncoder? encoder)
             return converted;
         }
 
+        private static string[] SplitFlagsEnum(string value)
+        {
+            // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
+            return value.Split(
+#if BUILDING_INBOX_LIBRARY
+                ValueSeparator
+#else
+                new string[] { ValueSeparator }, StringSplitOptions.None
+#endif
+                );
+        }
+
         internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            bool success = TryParseEnumQuick(ref reader, options, out T value, out string? _);
+            if (!success)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            return value;
+        }
+
+        private T ReadEnumFromString(ref Utf8JsonReader reader, JsonSerializerOptions options)
+        {
+            bool success = TryParseEnumQuick(ref reader, options, out T value, out string? enumString);
+
+            if (success)
+            {
+                return value;
+            }
+
+            if (_namingPolicy == null)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            if (enumString == null && reader.TokenType != JsonTokenType.Null)
+            {
+                enumString = reader.GetString();
+            }
+
+            if (enumString == null)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            Debug.Assert(_enumValueCache != null, "Enum value cache should be instantiated if a naming policy is specified.");
+
+            if (!enumString.Contains(ValueSeparator))
+            {
+                success = _enumValueCache.TryGetValue(enumString, out value);
+            }
+            else
+            {
```

Comment:
Following https://github.com/dotnet/runtime/pull/73348/files#r938652828, this could be written as
```C#
```suggestion
if (!enumValueCache.TryGetValue(enumString, out value) && enumString.Contains(ValueSeparator))
{
```
with the final result being cached if there still is capacity.